setuptools_rust/build.py

```diff
@@ -345,6 +345,26 @@ def install_extension(
                 os.makedirs(os.path.dirname(ext_path), exist_ok=True)
 
             log.info("Copying rust artifact from %s to %s", dylib_path, ext_path)
+
+            # We delete any existing library file before copying to avoid
+            # causing running processes that use it from segfaulting.
+            #
+            # This is because shared libraries are memory mapped into processes
+            # that use it, so modifying the shared library file (which is what
+            # `shutil.copyfile` does) means the next time a process calls into
+            # the shared library they get the updated library rather than the
+            # original one, causing confusion and segfaults.
+            #
+            # Deleting a memory mapped file, on the other hand, is correctly
+            # handled by the OS, i.e. a copy is kept around until the running
+            # process exits. Thus, deleting the existing library file and *then*
+            # copying the new one won't change the existing memory mapped file,
+            # and so is safe to do.
+            try:
+                os.remove(ext_path)
+            except FileNotFoundError:
+                pass
+
             shutil.copyfile(dylib_path, ext_path)
```

Comment:
If we try to make this more robust, wouldn't an atomic replacement be better here, e.g. move to `".~" + ext_path` and then move `".~" + ext_path` to `ext_path` to atomically replace the name with the copy of the library? Otherwise it seems possible that processes that try to start and map a partially copied library might fail as well.

Reply:
Yeah, we could do this as a `os.rename(..)` after deleting the existing file. I don't follow why we would do two moves?

Reply:
As it is implemented now, there is a window after removing the existing library and until moving the new one in its place when there is no library there. This would be avoid by the copy-then-rename dance.

(I am also not sure whether moving the library is the right thing to do since it removes the build output. `os.rename` also cannot cross file system boundaries which is why copying the file into the target directory first is usually preferable.)

Reply:
> As it is implemented now, there is a window after removing the existing library and until moving the new one in its place when there is no library there. This would be avoid by the copy-then-rename dance.

I didn't think there was a way to atomically swap two files? 

>  (I am also not sure whether moving the library is the right thing to do since it removes the build output. os.rename also cannot cross file system boundaries which is why copying the file into the target directory first is usually preferable.)

Aha! Right, that makes sense, so the `shutil.copyfile` should become `shutil.copyfile` to same directory then a `rename` to the actual path? Though I think we still need to delete the existing library before the rename, as rename fails if the path already exists?

Reply:
> Though I think we still need to delete the existing library before the rename, as rename fails if the path already exists?

I am not totally sure on Windows, but on Unix-like systems this exactly the idiom to atomically replace the file. (It is not guaranteed by POSIX IIRC but basically all real Unix filesystems guarantee in their common configurations.)

Reply:
(Meaning one can use `rename` to overwrite a file without removing it first.)

Reply:
Just had a look at the Python docs and it does not work on Windows but it is required by POSIX nowadays: https://docs.python.org/3/library/os.html#os.rename

But they also recommend `os.replace` as the cross platform way of achieving exactly this: https://docs.python.org/3/library/os.html#os.replace

So I think copying to the target directory and then `os.replacing` without removing first would be the best approach.

Reply:
Huh, well TIL. Thank you very much! 

I don't know what I was reading when I was previously looking at the docs, but [`os.rename`](https://docs.python.org/3/library/os.html#os.rename) clearly states what's going on:

> On Windows, if dst exists a [FileExistsError](https://docs.python.org/3/library/exceptions.html#FileExistsError) is always raised.
>
> On Unix, if both are files, dst will be replaced silently if the user has permission. The operation may fail on some Unix flavors if src and dst are on different filesystems. If successful, the renaming will be an atomic operation (this is a POSIX requirement).

So I think on Windows we'd still need to do the `os.remove`? But equally, I'm not sure if Windows suffers the same problem, so we could just leave it as is?